### ISSUE TITLE
`Ingredient.fromJson` in a custom recipe class results in an Empty Tag.

### ISSUE THREAD

**Issue by HaydenBelanger** [MinecraftForge/MinecraftForge#8498]
**Minecraft Version:** 1.18.2

**Forge Version:** 40.0.3

**Logs:** N/A

**Description of issue:**

Certain tags used within certain recipe Serializers result in an Empty Tag once in-world.

**Steps to Reproduce:**
 1. Use `Ingredient.fromJson` within any non-vanilla recipe Serializer to try to create an Ingredient from JSON.
 2. Enter game and use JEI/recipe book to view that the game shows an Empty Tag. Recipe is non-processable by expected functions.
 3. Arbitrarily, tag may fill itself at some point during operation until world is unloaded again, possibly related to /reload.

*JEI used to demonstrate*
![View with JEI](https://user-images.githubusercontent.com/7004666/156865539-ac841770-8582-4551-b82e-ee79c72e31fb.PNG)

[Link to image serializer](https://github.com/HaydenBelanger/assemblylinemachines/blob/530fdbeeea6c9d13fe57fdebb3ef65e6fb9bb4dc/src/main/java/me/haydenb/assemblylinemachines/crafting/BathCrafting.java#L213-L290)

[Link to image recipe JSON](https://github.com/HaydenBelanger/assemblylinemachines/blob/530fdbeeea6c9d13fe57fdebb3ef65e6fb9bb4dc/src/main/resources/data/assemblylinemachines/recipes/enchanted_book/looting.json)

**Comment by SizableShrimp** [comment]
Can you provide the specific ingredient JSON for clarity?

**Comment by HaydenBelanger** [comment]
Edited to include.

**Comment by pupnewfster** [comment]
As of 1.18.2 I don't believe vanilla's ingredient's in recipes (at least for tag ingredients) return valid results from getItems until the reload is complete. So your serializer needs to delay the consumption of this until tags are resolved. (The reason it works after a reload is that you are making use of the previous reload's data, which might even be out of date so should not be used)

A similar issue I believe exists with forge's code and how it initializes Ingredient.isSimple in the constructor, even though getItems no longer will provide an accurate result for that for tags during deserialization.

Edit: My best guess as to why they do this is so that they don't need to keep two sets of tags around during reloading and then ensure they are referencing the correct one, rather than just referencing the "key" and then not accessing it until the reload is complete and the tags have been updated.

**Comment by Eelt** [comment]
Just to add, this is the case even when not using `Ingredient.java`, [For example:](https://github.com/MikeTheShadow/VoluminousEnergy/blob/88ad3c89b79bfb9102833baf78525c85297951a9/src/main/java/com/veteam/voluminousenergy/recipe/CentrifugalAgitatorRecipe.java#L171-L173)
```
                TagKey<Fluid> tag = TagKey.create(Registry.FLUID_REGISTRY, fluidTagLocation);
                if (tag != null){
                    for (Holder<Fluid> fluidHolder : Registry.FLUID.getTagOrEmpty(tag)){ 
```
would also always be empty, as the `#getTagOrEmpty()` would return empty the first time around.

I'm guessing by pupnewfster's response is that a reload needs to be completed before a registry could return a populated list of holders, and the second time it passes a populated list that you can iterate through because the last one was cached? Even then, how would you even go around "delaying" the serializer to wait for the tags to be populated in order to build a valid recipe? Especially important if you're trying to avoid using ingredients for fluids, or I guess maybe even trying to make a custom fluid ingredient?

Edit: the above code would correspond with 269-271 in `TagValue#getItems();` which I guess was already established to not return valid items. So the real question would more or less be delaying serialization until tags are resolved...

**Comment by Eelt** [comment]
I think the scope of this issue may potentially be much larger, not loading data from datapacks properly in general. Loot doesn't generate in chests at all in `40.0.3`. My custom tools which also use tags appear to have a potentially similar issue with the tags being empty (and `blockState#is()` checks failing because of this on `getDestroySpeed`)...

**Comment by pupnewfster** [comment]
> I'm guessing by pupnewfster's response is that a reload needs to be completed before a registry could return a populated list of holders, and the second time it passes a populated list that you can iterate through because the last one was cached? Even then, how would you even go around "delaying" the serializer to wait for the tags to be populated in order to build a valid recipe? Especially important if you're trying to avoid using ingredients for fluids, or I guess maybe even trying to make a custom fluid ingredient?

You don't delay the serializer, you delay resolving the tag to the first time it is queried as part of the recipe in the same way vanilla's ingredients only track the TagKey now instead of the actual value of the tag and then caches it in the call to ingredient getItems.

If there are cases where some elements are never valid and you want to make sure they don't get added, then you could just filter them out during the initial call that ends up caching the values.

**Comment by malte0811** [comment]
There's a second issue in Forge caused by this change, [`TagEmptyCondition`](https://github.com/MinecraftForge/MinecraftForge/blob/4311edd4cd6940656f4a14e4dc5d9e6c2e51247d/src/main/java/net/minecraftforge/common/crafting/conditions/TagEmptyCondition.java#L48) no longer works as intended (reports all tags as empty). In this case it isn't possible to just delay the tag element lookup since the condition has to be evaluated during recipe loading.

**Comment by HaydenBelanger** [comment]
As @pupnewfster said, I was able to get around this for the purposes of a basic recipe serializer ONLY by switching to using `Lazy<Ingredient>` and moving valid-ingredient logic to the `Lazy.of`. But my solution only really works as a patchwork and @malte0811 still has a valid issue that can't be circumvented as easily.

**Comment by Eelt** [comment]
Okay, more findings on this: On world loading I'm observing this in the console in `40.0.4`, which are all forge tags (click details to expand):

<details>

```
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/redstone as it is missing following references: #minecraft:redstone_ores (from )
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/copper as it is missing following references: #minecraft:copper_ores (from )
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/gold as it is missing following references: #minecraft:gold_ores (from )
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/diamond as it is missing following references: #minecraft:diamond_ores (from )
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/lapis as it is missing following references: #minecraft:lapis_ores (from )
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/iron as it is missing following references: #minecraft:iron_ores (from )
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/coal as it is missing following references: #minecraft:coal_ores (from )
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/emerald as it is missing following references: #minecraft:emerald_ores (from )
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores as it is missing following references: #forge:ores/coal (from ), 
	#forge:ores/copper (from ), 
	#forge:ores/diamond (from ), 
	#forge:ores/emerald (from ), 
	#forge:ores/gold (from ), 
	#forge:ores/iron (from ), 
	#forge:ores/lapis (from ), 
	#forge:ores/redstone (from )
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:fences/wooden as it is missing following references: #minecraft:wooden_fences (from )
[20:38:05] [Worker-Main-35/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:fences as it is missing following references: #forge:fences/wooden (from )
[20:38:05] [Worker-Main-45/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/redstone as it is missing following references: #minecraft:redstone_ores (from )
[20:38:05] [Worker-Main-45/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/diamond as it is missing following references: #minecraft:diamond_ores (from )
[20:38:05] [Worker-Main-45/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/coal as it is missing following references: #minecraft:coal_ores (from )
[20:38:05] [Worker-Main-45/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:fences/wooden as it is missing following references: #minecraft:wooden_fences (from )
[20:38:05] [Worker-Main-45/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/lapis as it is missing following references: #minecraft:lapis_ores (from )
[20:38:05] [Worker-Main-45/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores/emerald as it is missing following references: #minecraft:emerald_ores (from )
[20:38:05] [Worker-Main-45/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:ores as it is missing following references: #forge:ores/coal (from ), 
	#forge:ores/diamond (from ), 
	#forge:ores/emerald (from ), 
	#forge:ores/lapis (from ), 
	#forge:ores/redstone (from )
[20:38:05] [Worker-Main-45/ERROR] [minecraft/TagLoader]: Couldn't load tag forge:fences as it is missing following references: #forge:fences/wooden (from )
```
</details>
Secondly, using Lazy<Ingredient> objects DOES fix for my recipes only if the tag is added by my mod, I've noticed that tags from vanilla still do not work even with using lazy ingredients in your recipes. (I have not tried using tags from other mods yet).

Example 1: [Recipe JSON](https://github.com/MikeTheShadow/VoluminousEnergy/blob/9b35f7d3457d2e98ad3d5cbf589b18cfa4b7741f/src/main/resources/data/voluminousenergy/recipes/stirling/coal.json), [Recipe Class](https://github.com/MikeTheShadow/VoluminousEnergy/blob/9b35f7d3457d2e98ad3d5cbf589b18cfa4b7741f/src/main/java/com/veteam/voluminousenergy/recipe/StirlingGeneratorRecipe.java) 
Example 2: [Recipe JSON](https://github.com/MikeTheShadow/VoluminousEnergy/blob/9b35f7d3457d2e98ad3d5cbf589b18cfa4b7741f/src/main/resources/data/voluminousenergy/recipes/crushing/ores/vanilla/coal_ore.json), [Recipe Class](https://github.com/MikeTheShadow/VoluminousEnergy/blob/9b35f7d3457d2e98ad3d5cbf589b18cfa4b7741f/src/main/java/com/veteam/voluminousenergy/recipe/CrusherRecipe.java)

Basically, from what I can see, anything in the `minecraft` namespace continues to have empty tags despite using Lazy ingredients, unless I did something wrong/misunderstood what everyone else here on this thread meant by using Lazies for resolving tags. In fact, seems the logged Forge errors include the tags that I was trying to use (plus a few others that I wasn't) so could be an issue beyond just in recipes.

**Comment by Eelt** [comment]
I have to agree with @malte0811, Lazies is not the solution to this problem. I've spent more time to properly make my recipes lazy by essentially deferring all the logic inside of Lazies and for `toNetwork` and `fromNetwork` methods, basically just sending a flag to see if my recipe uses a tag, and if so, the client actually builds the recipe contents (fluids and items) inside the `fromNetwork` method, essentially only receiving a text component from the server with the `ResourceLocation` of the tag encoded as a string.

 I'd like to also add is the tags that are broken in the console output from my last comment, are actually broken in _Vanilla_ (see images in details):

<details> 
![image](https://user-images.githubusercontent.com/6241870/157276755-8c1939cc-e441-40f9-9592-0239612ffbaa.png)
![image](https://user-images.githubusercontent.com/6241870/157276816-e4986e13-d069-4eaa-86a1-3206d2d823bf.png)
</details>

So one is not even able to craft a pickaxe at the **vanilla crafting table**, because the tag `minecraft:stone_tool_materials` is empty... That's the _real_ issue here is the tags being broken (no amount of lazies can fix this; it does fix it for the tags that aren't broken though). I'm waiting on PR#8495 to see if that fixes the issues with vanilla and forge tags being broken.

**Comment by XFactHD** [comment]
> So one is not even able to craft a pickaxe at the vanilla crafting table, because the tag minecraft:stone_tool_materials is empty... That's the real issue here is the tags being broken

This sounds like you are testing on 40.0.0, you need to update to at least 40.0.2 for tags containing modded objects to work properly.

**Comment by Eelt** [comment]
I'm observing this on Forge version 40.0.12
![image](https://user-images.githubusercontent.com/6241870/157282470-a5dda99d-a06b-4359-8545-3b7bfa4a78a1.png)
Tags containing modded objects work fine for me, but certain vanilla and forge namespace tags are broken for me (see console output my my comment 3 days ago)

EDIT: this only happens in my MDK environment. Compiling my mod jar and loading it as an end-user is perfectly fine... So it's not an issue with my mod, but something in particular with my MDK environment.

**Comment by sciwhiz12** [comment]
So, the cause of this issue is that a recipe serializer is performing validation on recipe JSON inputs during serialization. 

As @pupnewfster notes, the items within an `Ingredient` _must not_ be resolved (through `getItems()`) during recipe serialization, as they may involve the resolution of the values of tags. The values of tags are only bound to their registries after `ReloadableServerResources` is created and made available to the server, which happens after all resource listeners have finished loading -- including the tag manager and the recipes manager.

To my understanding, the issue shown with JEI recognizing the tag as empty is because `Ingredient` caches the array of item stacks returned by `getItems()`[^1]. Because the tag was resolved earlier when its values were not available, the cache held by the `Ingredient` instance is empty, even after successive calls. The cache is never cleared and rebuilt normally because the only time the values can change mid-game is through a reload, which also means the `Ingredient`s will be recreated from scratch. 

As @pupnewfster further notes, resolutions after the initial load work only because the registry (on the server-side) does not reset the values of tags before loading. This means that those resolutions are using the values of the tags from the _previous_ load, rather than the current load. 

I view that as a bug, but not on Mojang's side: Forge should probably implement a patch to call `resetTags` on all registries before reloading and before closing a world to ensure that any resolution of tags after the initial load remains consistently faulty.

As a side-note: Minecraft itself does not run into this problem, as it performs no validation on recipe ingredients during deserialization.

Therefore, I will be **closing** this issue on the basis that this is due to faulty mod behavior that must adapt to the new circumstances of _never_ resolving a tag (and by extension, an `Ingredient`'s item list which may be dervied from a tag) during recipe deserialization.

---

@malte0811, the issue with the `forge:tag_empty` recipe condition will be tracked by #8509.

[^1]: To clarify: this is not a bug with JEI. It is a bug due to the early resolution of the tag-containing ingredient.
